controllers/authpolicy_status.go

```diff
@@ -71,27 +58,83 @@ func (r *AuthPolicyReconciler) reconcileStatus(ctx context.Context, ap *api.Auth
 	newStatus := &api.AuthPolicyStatus{
 		Conditions:         slices.Clone(ap.Status.Conditions),
 		ObservedGeneration: ap.Status.ObservedGeneration,
 	}
 
-	availableCond := r.acceptedCondition(ap, specErr, authConfigReady)
-
+	availableCond := r.acceptedCondition(ap, specErr)
 	meta.SetStatusCondition(&newStatus.Conditions, *availableCond)
 
+	// Do not set enforced condition if Accepted condition is false
+	if meta.IsStatusConditionFalse(newStatus.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)) {
+		return newStatus
+	}
+
+	enforcedCond := r.enforcedCondition(ctx, ap, targetNetworkObject)
+	meta.SetStatusCondition(&newStatus.Conditions, *enforcedCond)
+
 	return newStatus
 }
 
-func (r *AuthPolicyReconciler) acceptedCondition(policy common.KuadrantPolicy, specErr error, authConfigReady bool) *metav1.Condition {
-	cond := common.AcceptedCondition(policy, specErr)
+func (r *AuthPolicyReconciler) acceptedCondition(policy common.KuadrantPolicy, specErr error) *metav1.Condition {
+	return common.AcceptedCondition(policy, specErr)
+}
+
+// enforcedCondition checks if the provided AuthPolicy is enforced based on the status of the associated AuthConfig and Gateway.
+func (r *AuthPolicyReconciler) enforcedCondition(ctx context.Context, policy *api.AuthPolicy, targetNetworkObject client.Object) *metav1.Condition {
+	logger, _ := logr.FromContext(ctx)
+	authConfigReady, gwPolicyOverridden, err := r.checkAuthConfigAndGateway(ctx, policy)
+	if err != nil {
+		logger.Error(err, "Failed to check AuthConfig and Gateway")
+		return common.EnforcedCondition(policy, common.NewErrUnknown(policy.Kind(), err))
+	}
 
 	if !authConfigReady {
-		cond.Status = metav1.ConditionFalse
-		cond.Reason = "AuthSchemeNotReady"
-		cond.Message = "AuthScheme is not ready yet" // TODO(rahul): need to take care if status change is delayed.
+		logger.V(1).Info("AuthConfig is not ready")
+		return common.EnforcedCondition(policy, common.NewErrUnknown(policy.Kind(), errors.New("AuthScheme is not ready yet")))
+	}
+
+	if gwPolicyOverridden {
+		logger.V(1).Info("Gateway Policy is overridden")
+		return r.handleGatewayPolicyOverride(logger, policy, targetNetworkObject)
 	}
 
-	return cond
+	logger.V(1).Info("AuthPolicy is enforced")
+	return common.EnforcedCondition(policy, nil)
+}
+
+// checkAuthConfigAndGateway checks if the AuthConfig is ready and if the Gateway Policy is overridden.
+func (r *AuthPolicyReconciler) checkAuthConfigAndGateway(ctx context.Context, policy *api.AuthPolicy) (bool, bool, error) {
+	apKey := client.ObjectKeyFromObject(policy)
+	authConfigKey := client.ObjectKey{
+		Namespace: policy.Namespace,
+		Name:      authConfigName(apKey),
+	}
+	authConfig := &authorinoapi.AuthConfig{}
+	err := r.GetResource(ctx, authConfigKey, authConfig)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, false, fmt.Errorf("failed to get AuthConfig: %w", err)
+		}
+		return true, common.IsTargetRefGateway(policy.GetTargetRef()), nil
```

Comment:
If we cannot find the AuthConfig, then the AuthConfig is ready? Twisted as it sounds, I understand this is what we want. Maybe add a comment?

Reply:
This would happen when the AuthPolicy hasn't reconciled yet the AuthConfig, right?

Reply:
Yeah, not too happy with this logic either to detect that an AuthConfig is not required from an AuthPolicy in this overridden scenario 🤔 

Re-thinking this, ideally we should be able to infer this from the AuthPolicy struct from a value set which denotes this scenario here:
https://github.com/Kuadrant/kuadrant-operator/blob/133645e9080f55022eff7c67d22da68424bce04a/controllers/authpolicy_authconfig.go#L92-L95

Instead I'm thinking we can add an annotation to denote that a gateway AuthConfig is required or not `gwAuthConfigRequired: true|false` instead of this twisted logic and also to prevent getting this is code logic if we are actually waiting for Authconfig to be created 🤔 

What do you guys think ?

Reply:
> ideally we should be able to infer this from the AuthPolicy struct from a value set which denotes this scenario here:

Thing is that it's not the structure of the Gateway AuthPolicy itself that says that, but the presence of other AuthPolicies targeting the HTTPRoutes, as you know.

The LOC you pointed is indeed where this decision happens. Signaling (possibly by annotating, but not necessarily) from there, all the way to the step where we set the status, sounds a good idea to me.

> I'm thinking we can add an annotation to denote that a gateway AuthConfig is required or not 

Do you mean an annotation in the AuthPolicy CR? Would that be actually pushed to etcd or just to carry in-memory the information to the set the status later on?

In Authorino, we implemented an approach of writing to a shared memory space that the status controller reads from: https://github.com/Kuadrant/authorino/blob/744220a503dbf231f104652f16084bb51feef9a3/controllers/auth_config_controller.go#L83

> `gwAuthConfigRequired: true|false`

I'm not yet 100% convinced about the annotation plan, but if we do that, please let's use a name such as `kuadrant.io/authconfig-required` instead please.

Reply:
> Do you mean an annotation in the AuthPolicy CR? Would that be actually pushed to etcd or just to carry in-memory the information to the set the status later on?
> 

Yes, annotation in the AuthPolicy CR that is kept in memory since this information is implied in the AuthPolicy status once the `Overridden` reason is set. I don't think there is a need to hold this information again in the AuthPolicy CR via annotations unless it proves useful for end users / consumers of AuthPolicy CR to know that there is no AuthConfig created for an AuthPolicy. Although I feel there is no major harm in pushing this annotation to etcd either 🤔 Using annotations was just an easy way of passing this information down to the Enforced status logic

> In Authorino, we implemented an approach of writing to a shared memory space that the status controller reads from: https://github.com/Kuadrant/authorino/blob/744220a503dbf231f104652f16084bb51feef9a3/controllers/auth_config_controller.go#L83
> 
> > `gwAuthConfigRequired: true|false`
> 
> I'm not yet 100% convinced about the annotation plan, but if we do that, please let's use a name such as `kuadrant.io/authconfig-required` instead please.

Thanks ! I'll have a look at Authorino to see if something similar is the preferred approach 👍